### ISSUE TITLE
Document Viewer: reload and show page scrolled down to match the editor's text cursor line

### ISSUE THREAD

**Issue by tmarplatt** [vkbo/novelWriter#1367]
**The problem**

When previewing a long enough document, the document viewer always loads unscrolled, showing the beggining of the document. If the user is currently writing it and wants to preview a paragraph beyond the first few ones, they're forced to scroll down manually every time.

**The solution**

When the document currently being previewed is also the one open in the document editor, then the document viewer should try to present it at about the same page height/scroll height matching the cursor location in the editor.

**Comment by vkbo** [comment]
Firstly, the viewer is not intended to be used as preview, so parallel scrolling features have not been implemented in general. The idea is that the markup used is so simple that preview is not needed in most cases.

In addition, if the viewer reloads the same document, it should preserve the scroll bar position it had before the reload, so as long as you had already scrolled to that paragraph, it should scroll back again after a reload. No matter how you initiate the reload.

Since you're saying it's always starting at the top, I suppose what you mean here is that if you load a document into the viewer, when it wasn't _already_ loaded there, it should scroll to the cursor position in the editor if the same document is open there? That is not necessarily a difficult feature to implement, the viewer has the capability to scroll to a specific line in the text quite easily, but doing so assumes the intention of the writer. This is always a tricky thing to guess. How do we know that the writer intends to compare only that one paragraph? Or that using the viewer as preview is the intention at all?

**Comment by tmarplatt** [comment]
True. Thinking about it, there are three things that, piled together, confused me. Disentangling them:

- What I really want is a new View Document At Cursor Position command. A shortcut, for keyboard based nW usage. Using the viewer to preview is a legitimate use case for certain markup (e.g. text alignment) and in a general sense (appraisal of both form and content of a work).
- The "Reload the document" button is a feature of its own, but View Document uses Ctrl+R (a shortcut which commonly stands for _reload page/screen/something_). I was using Ctrl+R as if it stood for Reload...
- "Reload the document" should be its own command/shortcut. And it should in fact reload the viewer at the same absolute scroll height (provided the viewer can scroll that far after rendering the text). I believe this should be standard behaviour for such a feature?

**Comment by vkbo** [comment]
> True. Thinking about it, there are three things that, piled together, confused me. Disentangling them:
> 
> * What I really want is a new View Document At Cursor Position command. A shortcut, for keyboard based nW usage. Using the viewer to preview is a legitimate use case for certain markup (e.g. text alignment) and in a general sense (appraisal of both form and content of a work).
> * The "Reload the document" button is a feature of its own, but View Document uses Ctrl+R (a shortcut which commonly stands for _reload page/screen/something_). I was using Ctrl+R as if it stood for Reload...

It does not stand for reload. It initially stood for "pReview", as `Ctrl+P` is standard for Print. The preview feature was requested by my colleague who I originally sketched out novelWriter with in the beginning. It was later generalised to a pure "View Document" functionality as it is labelled now. In the project tree, it loads the selected document in the viewer, and if the editor has focus, the one you're editing.

After thinking a bit more about it, I think that there is no other use for `Ctrl+R` from the editor than your scenario, so I'm thinking that it wouldn't be too far fetched to assume scrolling to the cursor position is better than not doing that as long as the editor has focus when you press it.

> * "Reload the document" should be its own command/shortcut. And it should in fact reload the viewer at the same absolute scroll height (provided the viewer can scroll that far after rendering the text). I believe this should be standard behaviour for such a feature?

The only actual case for a reload button in the viewer is the one case when the document is being edited, so that is already covered. The actual button in the viewer's header connects to the same logic that `Ctrl+R` does. The load document functionality in the viewer will detect any attempt to reload the current document, and preserve the scroll bar position in such cases, as mentioned in my previous response.

I don't think there is anything that needs changing on this side of it. The change should be when it's initiated from the editor.
